S2EExecutor: onStateForkDecide signal incorrectly emitted

adrianherrera opened the pull request:
The `onStateForkDecide` signal is spuriously emitted. The reason is as follows:

KLEE will call `S2EExecutor::fork` regardless of the branch's condition being a symbolic expression or concrete value. It is not until `klee::Executor::concolicFork` and `klee::Executor::fork` that the condition is checked and a fork performed **only** if the condition is **not** a `klee::ConstantExpr`.

However, the `onStateForkDecide` signal is emitted **before** this check for a `klee::ConstantExpr` is actually made. This signal is emitted unconditionally, which means that it will occur for every branch instruction in LLVM code, regardless of whether it depends on a symbolic expression or not.

For example, this can cause the following problem. If the program counter is checked in the `onStateForkDecide` signal handler, it may show addresses that correspond to instructions that are not necessarily branch instructions. This is because functions like `__ldb_mmu` (which get inserted into the generated LLVM code when performing a memory read) also contain conditional branches that have nothing to do with symbolic values. For example, one check is that `*g_sqi.events.before_memory_access_signals_count != NULL`. This branch condition has no symbolic expressions, however `S2EExecutor::fork` will still be called and the `onStateForkDecide` signal will still be emitted. This may lead to incorrect state fork decisions (e.g. if you are basing your decision on the program counter).

To fix this problem we restrict the `onStateForkDecide` signal so that it is only emitted when the branching condition is based on a non-constant symbolic expression.

vitalych commented:
I agree it doesn't make sense to emit the signal if the condition is concrete, as there is nothing to decide there. In general however, a symbolic expression does not guarantee that there will be a fork later (if the constraint solver decides otherwise). So you will still have that onForkDecide (on potentially any instruction, not just a jump) called without necessarily forking later.

The program counter however must be correct, in the sense that it will be the one of the instruction currently being executed. So it's still possible to take decisions based on it. Just don't assume that it will be a branch instruction.

I think all these explanations are very useful and must go in the code.

vitalych reviewed (changes requested):
see notes

Epeius commented:
I think the purpose of Adrian's code is to prevent emitting this signal multiple times for each execution of an instruction, not to avoid emitting signal for concrete conditions. While actually this code will get rid of all the concrete conditions, and only focus on symbolic conditions.

vitalych commented:
The signal can indeed be emitted multiple times per instruction. I think it's fine. Perhaps you'd like to have a signal fired once to disable/enable any subsequent forks on that instruction?

Epeius commented:
Not per instruction but per each step execution (maybe I didn't explain correctly), we found that the signal will emit multiple times for an instruction even the instruction is only executed once.

adrianherrera commented:
Ok, the more I think about it the more I think that it is not a bug and just a misunderstanding on how this signal should work. **Warning**: long message ahead :)

> In general however, a symbolic expression does not guarantee that there will be a fork later (if the constraint solver decides otherwise). 

This makes sense to me. In fact it is obvious now that it has been pointed out :)

> Just don't assume that it will be a branch instruction.

~~This I don't understand. Where else would we fork but on a branch or switch instruction?~~ I understand now.

Maybe I should illustrate with my use case, and you can tell me if the behaviour or my understanding is incorrect.

I have the following code in my SUT:

```
0x771BD82B mov     eax, large fs:18h
0x771BD831 mov     eax, [eax+30h]
0x771BD834 cmp     byte ptr [eax+2], 0
0x771BD838 jnz     loc_771FD942
```

The data at `[eax+2]` is concolic, so I expect a fork at `0x771BD838`.

At runtime the following LLVM code is generated:

```
  ; Update the program counter so that it points to the cmp instruction
store i64 0x771BD834, i64* %2
  ; Get [eax+2]
%6 = getelementptr %struct.CPUX86State, %struct.CPUX86State* %state, i32 0, i32 0, i32 0
%rax_v = load i64, i64* %6
%tmp2_v = add i64 %rax_v, 2
%tmp2_v1 = and i64 %tmp2_v, 0xFFFFFFFF
%7 = call i8 @__ldb_mmu(i64 %tmp2_v1, i32 1)

  ; __ldb_mmu
%3 = load i32*, i32** getelementptr inbounds (%struct.se_libcpu_interface_t, %struct.se_libcpu_interface_t* @g_sqi, i64 0, i32 9, i32 0), align 8
%4 = load i32, i32* %3, align 4
%5 = icmp eq i32 %4, 0
br i1 %5, label %7, label %6
  ; onStateForkDecide emitted. Nothing to do with the data in %tmp2_v1
%8 = load i32*, i32** getelementptr inbounds (%struct.se_libcpu_interface_t, %struct.se_libcpu_interface_t* @g_sqi, i64 0, i32 1, i32 1), align 8
%9 = load i32, i32* %8, align 4
%10 = icmp eq i32 %9, 0
br i1 %10, label %13, label %11
  ; onStateForkDecide emitted. Nothing to do with the data in %tmp2_v1

  ; onStateForkDecide gets emitted another 3 times in __ldb_mmu
```

As we can see in the generated LLVM, `onStateForkDecide` is emitted 5 times for the instruction at `0x771BD834`. When I inspect the PC in S2E, I see that it is `0x771BD834`. This corresponds to a `cmp` instruction.

There are two things that didn't seem logical to me when I first encountered this:
1. Why is the signal emitted 5 times for the same instruction (as Bin noted)?
2. Why are we "forking" on a `cmp` instruction? Shouldn't the fork occur at the `jnz`? (In reality there is no fork because all of the conditions in the LLVM `icmp` instructions are `klee::ConstantExpr` expressions, and it does fork at the `jnz`)

This is why I went down the rabbit hole of "the PC must not be updating correctly". In the end it has nothing to do with the code in the SUT itself, but it is because of code that gets inserted by the LLVM code generator (in this case, `__ldb_mmu`). This is very confusing.

So, is this correct behaviour? If it is correct behaviour, I will replace my code changes with comments/documentation that explains this for others so they do not fall down the same hole.

adrianherrera commented:
Another thought that I had on the walk to work this morning; can we push the `onStateForkDecide` signal deeper into the KLEE code? That way we will have more knowledge of what KLEE will actually do (e.g. it may have already decided not to fork because the condition is a `klee::ConstantExpr`).

For example, https://github.com/S2E/klee/blob/master/lib/Core/Executor.cpp#L532 has the same logic as the code in this PR. Does it make sense to add a `notifyStateForkDecide` method that gets called at L536 and emits the `onStateForkDecide` signal? (Similar logic would have to be added to `klee::fork`.)

vitalych commented:
You describe the correct behavior. If a memory access is symbolic, S2E may be forced to fork and concretize, possibly several times, depending on the complexity of the memory access. That's why you may have forks for other instructions besides x86 branches. In general, any instruction that dereferences symbolic memory pointer can potentially fork.

Your fix is good, it filters the case where data is concrete, so it reduces noise. You should keep it this way. Just summarize this thread in a comment.

vitalych commented:
As for pushing the decision deeper into klee, not sure about it. It's already pretty close already, right before calling either fork or concolic fork. You'd have to duplicate the call twice if you push it.
https://github.com/S2E/libs2ecore/blob/master/src/S2EExecutor.cpp#L2085

Commit:
S2EExecutor: onStateForkDecide signal incorrectly emitted

KLEE will call S2EExecutor::fork regardless of the branch's condition being a
symbolic expression or concrete value. It is not until
klee::Executor::concolicFork and klee::Executor::fork that the condition is
checked and a fork performed **only** if the condition is **not** a
klee::ConstantExpr.

However, the onStateForkDecide signal is emitted **before** this check for a
klee::ConstantExpr is actually made. This signal is emitted unconditionally,
which means that it will occur for every branch instruction in LLVM code,
regardless of whether it depends on a symbolic expression or not.

For example, this can cause the following problem. If the program counter is
checked in the onStateForkDecide signal handler, it may show addresses that
correspond to instructions that are not necessarily branch instructions. This
is because functions like __ldb_mmu (which get inserted into the generated
LLVM code when performing a memory read) also contain conditional branches that
have nothing to do with symbolic values. For example, one check is that
`*g_sqi.events.before_memory_access_signals_count != NULL`. This branch
condition has no symbolic expressions, however S2EExecutor::fork will still be
called and the onStateForkDecide signal will still be emitted. This may lead to
incorrect state fork decisions (e.g. if you are basing your decision on the
program counter).

To fix this problem we restrict the onStateForkDecide signal so that it is only
emitted when the branching condition is based on a non-constant symbolic
expression.

Commit:
CorePlugin: updated comments/documentation

adrianherrera commented:
Ok, I've updated the comments in the CorePlugin header file. Please have a read (in particular the onStateForkDecide signal) and let me know if it is accurate.

Thanks!

vitalych commented on include/s2e/CorePlugin.h:
```diff
@@ -122,240 +133,321 @@ class CorePlugin : public Plugin {
+    ///
+    /// Note that this signal may be emitted when executing program instructions that are \b not necessarily branch
+    /// instructions. For example, when dereferencing symbolic memory a number of helper functions are called (see
+    /// \c TCGLLVMContextPrivate::initialzeHelpers) which may also cause fork branches. Depending on the: program
```
the: ?

adrianherrera replied:
Woops, typo. I'll fix
